<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro"  name="tennis_ball">

  <!-- Material Definition -->
  <material name="yellow">
    <color rgba="1 1 0 1"/>
  </material>

  <material name="black">
    <color rgba="0 0 0 1"/>
  </material>

  <link name="world">
  </link>

  <link name="base_link">
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <visual>
      <geometry>
        <box size="0.6 0.01 0.01"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.6 0.01 0.01"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.008333333333" ixy="0" ixz="0" iyy="0.1541666667" iyz="0" izz="0.1541666667"/>
    </inertial>
  </link>

  <!-- The gantry robot -->
  <link name="x_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.6 0.01"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.6 0.01"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.01541666667" ixy="0" ixz="0" iyy="0.0008333333333" iyz="0" izz="0.01541666667"/>
    </inertial>
  </link>
  
  
  <link name="y_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.6"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.6"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.01541666667" ixy="0" ixz="0" iyy="0.01541666667" iyz="0" izz="0.0008333333333"/>
    </inertial>
  </link>

  
  <link name="z_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="black"/>
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.1"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0008333333333" ixy="0" ixz="0" iyy="0.0008333333333" iyz="0" izz="0.0008333333333"/>
    </inertial>
  </link>
  
  <!-- Link Definition -->
  <link name="ball">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.033"/>
      </geometry>
      <material name="yellow"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <sphere radius="0.033"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
    </inertial>
  </link>
  
  <!-- Joints  -->
  <joint name="world_to_base_link_joint" type="fixed">
      <origin xyz="0.285 0 0.005" rpy="0 0 0"/>
      <parent link="world"/>
      <child link="base_link"/>
  </joint>

  <joint name="x_joint" type="prismatic">
      <origin xyz="0 0.305 0" rpy="0 0 0"/>
      <parent link="base_link"/>
      <child link="x_link"/>
      <axis xyz="1 0 0"/>
      <limit effort="50" velocity="0.01" lower="-0.3" upper="0.3"/>
      <dynamics damping="100.0" friction="100.0"/> 
  </joint>

  <joint name="y_joint" type="prismatic">
      <origin xyz="0 0 0.305" rpy="0 0 0"/>
      <parent link="x_link"/>
      <child link="y_link"/>
      <axis xyz="0 1 0"/>
      <limit effort="50" velocity="0.01" lower="-0.3" upper="0.3"/>
      <dynamics damping="100.0" friction="100.0"/> 
  </joint>

  <joint name="z_joint" type="prismatic">
      <origin xyz="0 -0.01 0" rpy="0 0 0"/>
      <parent link="y_link"/>
      <child link="z_link"/>
      <axis xyz="0 0 1"/>
      <limit effort="50" velocity="0.01" lower="-0.3" upper="0.3"/>
      <dynamics damping="100.0" friction="100.0"/> 
  </joint>

  <joint name="fixed_joint" type="fixed">
      <origin xyz="0 -0.033 0" rpy="0 0 0"/>
      <parent link="z_link"/>
      <child link="ball"/>
  </joint>

  <xacro:include filename="gazebo_plugin.xacro" />


</robot>


<!-- ros2 topic pub -1 /set_joint_trajectory trajectory_msgs/msg/JointTrajectory  '{header: {frame_id: world}, joint_names: [x_joint, y_joint, z_joint], points: [{positions: {0.3,0.3,0.3}, velocities: {0.02,0.01,0.01}}}]} -->